lib/engine/step/g_1860/tracker.rb

```diff
@@ -30,149 +32,51 @@ def lay_tile_action(action)
-              !@game.loading &&
-              !@game.graph.reachable_hexes(spender)[hex]
-
-            free = ability.free
-            discount = ability.discount
-            extra_cost += ability.cost
-          end
-
-          @game.abilities(entity, :teleport) do |ability, _|
-            ability.use! if ability.hexes.include?(hex.id) && ability.tiles.include?(tile.name)
-          end
-
-          terrain = old_tile.terrain
-          cost =
-            if free
-              # call for the side effect of deleting a completed border cost
-              border_cost(tile, entity)
-
-              extra_cost
-            else
-              border, border_types = border_cost(tile, entity)
-              terrain += border_types if border.positive?
-              @game.upgrade_cost(old_tile, hex, entity) + border + extra_cost - discount
-            end
-
-          if @game.insolvent?(spender) && cost.positive?
-            raise GameError, "#{spender.id} cannot pay for a tile when insolvent"
-          end
+        def save_old(old_tile, entity)
+          return if @game.loading || !entity.operator?
 
-          spender.spend(cost, @game.bank) if cost.positive?
-
-          cities = tile.cities
-          if old_tile.paths.empty? &&
-              tile.paths.any? &&
-              cities.size > 1 &&
-              cities.flat_map(&:tokens).any?
-            token = cities.flat_map(&:tokens).find(&:itself)
-            @round.pending_tokens << {
-              entity: entity,
-              hexes: [action.hex],
-              token: token,
-            }
-
-            token.remove!
-          end
-          @log << "#{spender.name}"\
-            "#{cost.zero? ? '' : " spends #{@game.format_currency(cost)} and"}"\
-            " lays tile ##{tile.name}"\
-            " with rotation #{rotation} on #{hex.name}"\
-            "#{tile.location_name.to_s.empty? ? '' : " (#{tile.location_name})"}"
-
-          return unless terrain.any?
-
-          @game.all_companies_with_ability(:tile_income) do |company, ability|
-            if terrain.include?(ability.terrain) && (!ability.owner_only || company.owner == entity)
-              # If multiple borders are connected bonus counts each individually
-              income = ability.income * terrain.find_all { |t| t == ability.terrain }.size
-              @game.bank.spend(income, company.owner)
-              @log << "#{company.owner.name} earns #{@game.format_currency(income)}"\
-                " for the #{ability.terrain} tile built by #{company.name}"
-            end
-          end
+          @tr_distance = @game.biggest_train_distance(entity)
+          @saved_revenues = if old_tile.color == :white
+                              []
+                            else
+                              old_tile.nodes.select { |n| reachable_node?(entity, n, @tr_distance) }
+                                .map(&:max_revenue).sort
+                            end
```

Comment:
It might read easier if you split this out to a function

```
@saved_revenues = old_revenues(entity, old_tile)

function old_revenues(entity, old_tile)
  return [] if old_tile.color == :white

  old_tile.nodes.select { |n| reachable_node?(entity, n, @tr_distance) }
                          .map(&:max_revenue).sort
end
```

I didn't make any change here, but you might consider combining the map and select.

Reply:
I can break it out into a function, but I'll have to not do the @tr_distance calculation or save.

What do you mean "combining the map and select"? Aren't they already combined?

Reply:
I'm changing my approach a little bit. I found a way to not modify the base Tracker at all.